README.txt
Open Source Overview
============================
INFLO-Aimsun allows connecting the INFLO-SIM application with Aimsun for the development of AMS (Analysis Modeling and Simulation) Testbeds.

License information
-------------------
Licensed under the Aimsun End-User License Agreement. You may not use this code except in compliance with the License. You can request a copy of the License by writing to info@aimsun.com
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

System Requirements
-------------------------
INFLO-Aimsun requires an Aimsun platformSDK license to be compiled, and an Aimsun license to be executed. The hardware requirements for the use of this application are the same as those for Aimsun alone; see https://www.aimsun.com/aimsun/tech-specs/ for details.

Aimsun runs on Windows, Linux, or Mac, but INFLO-SIM is only available for Windows; therefore, even if the interface could be compiled for other operating systems, the only practical option is Windows.

Documentation
-------------
Compile the provided code to produce a .dll. Copy the library and the provided .xml in the folder where custom plugins are stored (Aimsun 8.0\plugins).Setup the Aimsun document file	- Create the subpath along which data will be collected	- Set the direction as the External ID of the subpath	- Go to Types > GKObject > Route > Subpath.	- Add GKSubpath::INFLO::IncreasingDirection, Data Type: String	- Go to Table View and display Subpath. Set the direction of the subpath in the Increasing Direction attribute	- Create a new script, set Object Type to Subpath, and link it to the provided .py script	- Update the path of the database in line 167	- Execute the INFLO database script on the subpath; the log should display “done” if the execution is successful	- In the INFLO API dialogue, set the subpath ID, the vehicle ID’s for connected vehicles and INFLO sleep time to 10 seconds minimum. Anything less will cut off communication between INFLO, Aimsun and the database too quickly. Set the paths to syncfile (“folderlocation”/Data.txt) and database (“folderlocation”/INFLOdatabase.accbd)	- Change the Scenario Demand according to the market penetration rate required
Setup INFLO-SIM	- Go to INFLOApps > INFLOApps > bin > Release > Config > INFLOConfig.xml	- Update the path in lines 5, 18 and 19 which connect to the database, the CV Data directory and TSS Data Directory, respectively.	- Update lines 22 to 28 accordingly:		o Roadway ID = Subpath ID		o EndMM can be found in the attributes of the last section of the subpath    (NB = 21.1, SB = 21.9)		o RecurringCongestionMMLocation is recommended to be set equal to EndMM
Run a simulation	- Open INFLOApp and set Syncfile in the INFLO Settings tab	- Press Start in the INFLOApp before starting a simulation		o The INFLOApp log should display messages about speed harmonization for every 20 seconds of simulation time.  This will cause the simulation to pause for a few moments to update the database and communicate with INFLO.		o One way to check if the App is pointing to the correct folder is to look at the information in the tab INFLO Configuration
Web sites
---------
The INFLO-Aimsun software is distributed through the USDOT's JPO Open Source Application Development Portal (OSADP)
http://itsforge.net/ 


RELEASE-NOTES.txt
- Release note identifier
INFLO-Aimsun interface, April 2017, version 1.0

- What's New
First release.

- System Requirements
Aimsun 8.0
Aimsun 8.0 platformSDK

- Features and changes
First release.

- Outstanding issues
None.

- Installation guide
Obtain from TSS-Transport Simulation Systems (info@aimsun.com) a license of the platfromSDK.
Compile the supplied source code.
Copy the compiled library and the provided .xml in the folder where Aimsun plugins are stored (Aimsun 8.0\plugins).

- Known Issues
None.

- Troubleshooting
support@aimsun.com

- FAQ

- Other Resources and Links
www.aimsun.com

INFLOPlugin.cxx

#include "INFLOPlugin.h"
#include "INFLODescription.h"
#include "INFLOExtension.h"
#include "GKSystem.h"
#include "GKModel.h"
#include "GKGeoObject.h"
#include "GKSignalRouter.h"

//---- INFLOPluginFactory -----------------------------------------------------------

GKPlugin * INFLOPluginFactory()
{
    return new INFLOPlugin();
}

//---- AimsunExtTest ---------------------------------------------------------

INFLOPlugin::INFLOPlugin() : GKPlugin()
{
    name = tr( "INFLO Extension" );
}

INFLOPlugin::~INFLOPlugin()
{
}

void INFLOPlugin::initGui(GGui * /*gui*/ )
{
}

QString INFLOPlugin::verify()
{
	QString		res;

#ifdef _ANG_DEMO
	isDemo();
#endif
	return res;
}

void INFLOPlugin::install()
{
    GKSystem::getSystem().registerAimsunExtension( new INFLODescription() );
}

void INFLOPlugin::docOpened( GGui * /*gui*/, GKModel * /*model*/ )
{	
}



INFLOExtension.h
//-*-Mode: C++;-*-
#ifndef _INFLOExtension_h_
#define _INFLOExtension_h_

#include "GKAimsunExtension.h"

#include "INFLOUtil.h"
#include "SSNetwork.h"
#include <QMap>
#include <QVector>

//---- INFLOExtension -----------------------------------------------------------

class GKDetector;
class GKModel;
class GKSection;
class GKSubPath;
class ARandomGenerator;


class INFLOEXPORT INFLOExtension : public GKAimsunExtension
{
public:
	INFLOExtension();
	~INFLOExtension();

	void init();
	void finish();
	void postManage(double time);

	struct MileMarkerInfo{
		GKSection * section;
		int segment;
		double startMM;
	};

	struct DetectorInfo{
		int detectorId;
		double realTime;
		double volume;
		double occupancy;
		double speed;
	};

	struct VehicleInfo{
		int idVeh;
		double speed;
		int idSection;
		double position;
		bool queued;
		double mileMarker;
	};
	struct SpeedActionInfo{
		int vehPos;
		int idSection;
		int laneId;
		int segment;
		double speed;
		double position;
		bool queued;
		double mileMarker;
		double compliance;
	};
	struct SpeedInfo{
		double speed;
		double endMM;
	};

	class DetectorStats{
	public:
		DetectorStats(){
			speed.reset();
			occupancy = 0;
			counts = 0;
		}

		SSStatisticOutput speed;
		double occupancy;
		int counts;
		int idDetector;
		QVector< GKDetector* > detectors;
	};

private:

	int             mInterval;
	int             INFLOFreq;
	int             INFLOSleep;
	int			 mCacc1;
	int             mCacc2;
	double       mNextInfloStats;
	double       mCurrentTime;
	double       mLatency, mLatencyMin, mLatencyMax;
	double       mPackageLoss, mPackageLossMin, mPackageLossMax;
	QString      mDbLocation;
	QString      mSyncFile;
	int             mNbPackageDropped;
	bool            mApplySpeeds;

	ARandomGenerator *mRandomBall;

	GKModel 		*mModel;
	GKSubPath       *mSubpath;

	QMap< int, double >                 mSectionMileMarkers;
	QMap< double, MileMarkerInfo >      mMileMarkerInfo;

	QMap< int, DetectorStats* >         mDetectors;
	QMap< int, QVector< void* > >      mActions;

	QMap< double, QVector<SpeedActionInfo> > mDelayedActions; //the key is in timeSta

	void readDescriptionValues();
	void mapSubpathMileMarkers();
	void doTheManage(double timeSta);
	void manageDetectors();
	void manageSpeedAction( double timeSta );
	QVector<DetectorInfo> collectDetectorStats() const;
	QVector<VehicleInfo> collectVehicleStats();
	QVector< SpeedInfo > getSpeed() const;
	MileMarkerInfo findMileMarkerInfo( double endMileMarker ) const;
	void writeData( const QVector<DetectorInfo> & detectorsInfo, const QVector<VehicleInfo>& vehiclesInfo );
	void updateSD(const QVector< SpeedInfo >& speedData, double timeSta);
	double findSpeedToApply( const QVector< INFLOExtension::SpeedInfo >& speedData, double mileMarkerFrom, double mileMarkerTo );
	bool packageDropped() const;
	double latencyApplied() const;
	void removeActionsFromSection( int sectionId );
};


#endif


INFLOExtension.cxx

#include "INFLOExtension.h"
#include "INFLODescription.h"

#include "AKIAction.h"
#include "AKIDetector.h"
#include "AKIInfVeh.h"
#include "AKISimul.h"

#include "GKColumn.h"
#include "GKCommon.h"
#include "GKConstants.h"
#include "GKDBUtil.h"
#include "GKDetector.h"
#include "GKExperiment.h"
#include "GKModel.h"
#include "GKReplication.h"
#include "GKScenario.h"
#include "GKSection.h"
#include "GKSubPath.h"
#include "GKType.h"

#include "FRand.h"

#include <QDebug>
#include <QSqlError>

#include <Windows.h>

double toMiles = 0.000621371;
double kmToMile = 0.621371;
double feetToMile = 0.000189394;

INFLOExtension::INFLOExtension() : GKAimsunExtension()
{
	mLatency = 0;
	mPackageLoss = 0;
	mModel = NULL;
	mNbPackageDropped = 0;
	mApplySpeeds = true;
}

INFLOExtension::~INFLOExtension()
{

}

void INFLOExtension::readDescriptionValues()
{
	QDomNode				firstNode;
	GKPreferencesValues values;
	QDomDocument		doc( "parameters" );

	doc.setContent( parameters );
	values.setDefinition( ( ( INFLODescription* )description )->getPreferencesDefinition() );
	firstNode = doc.firstChild();
	values.restore( firstNode );

	int subpathId = values.getValue( "INFLODescription::Values::SubpathId" ).toInt();
	mModel = replication->getModel();
	mSubpath = dynamic_cast<GKSubPath*>( mModel->getCatalog().find( subpathId ) );
	INFLOSleep = values.getValue( "INFLODescription::Values::Sleep" ).toInt();
	mDbLocation = values.getValue( "INFLODescription::Values::InfloDB" );
	mSyncFile = values.getValue( "INFLODescription::Values::SynchFile" );
	mCacc1 = values.getValue("INFLODescription::Values::Cacc1").toInt();
	mCacc2 = values.getValue("INFLODescription::Values::Cacc2").toInt();

	mLatencyMin = values.getValue( "INFLODescription::Values::LatencyMin").toDouble();
	mLatencyMax = values.getValue( "INFLODescription::Values::LatencyMax").toDouble();

	if( GK::AreEqual( mLatencyMin, mLatencyMax ) == true ){
		mLatency = mLatencyMin;
	}else{
		mLatency = mRandomBall->uniform( mLatencyMin, mLatencyMax );
	}

	mPackageLossMin = values.getValue( "INFLODescription::Values::PackageLossMin").toDouble() / 100.0;
	mPackageLossMax = values.getValue( "INFLODescription::Values::PackageLossMax" ).toDouble() / 100.0;

	mApplySpeeds = values.getValue("INFLODescription::Values::NotApplyInfloSpeeds").toInt() == 0;

	if( GK::AreEqual( mPackageLossMin, mPackageLossMax ) == true ){
		mPackageLoss = mPackageLossMin;
	}else{
		mPackageLoss = mRandomBall->uniform( mPackageLossMin, mPackageLossMax );
	}
}

void INFLOExtension::init()
{
	mRandomBall = new ARandomGenerator( A2KModel::getModel().SemillaAPIExtensions );
	readDescriptionValues();
	mapSubpathMileMarkers();
	INFLOFreq = 20;

	mNextInfloStats = INFLOFreq * 1000;
	mCurrentTime = 0;
	mInterval = 0;
}

void INFLOExtension::mapSubpathMileMarkers()
{
	GKColumn * startMileMarkerColumn = mModel->getColumn( "GKSection::MileMarker");

	foreach( GKSection * section, mSubpath->getRoute() ){
		double startMileMarker = section->getDataValueDouble( startMileMarkerColumn );
		mSectionMileMarkers[section->getId()] = startMileMarker;
		int nbSegments = section->getNbSegments();
		double segmentLength = 0;
		for( int index = 0; index < nbSegments; index++ ){
			INFLOExtension::MileMarkerInfo info;
			info.section = section;
			info.segment = index+1;
			info.startMM = startMileMarker+segmentLength;
			segmentLength += section->getSegmentLength( index ) * toMiles;
			mMileMarkerInfo[startMileMarker+segmentLength] = info;
			qDebug( "%f -> section %d segment %d", startMileMarker+segmentLength, info.section->getId(), index+1 );
		}

		GKGeoObjects* topObjects = section->getTopObjects();
		if( topObjects ){
			foreach( GKGeoObject* geoobject, *topObjects ){
				GKDetector * detector = dynamic_cast<GKDetector*>( geoobject );
				if( detector && detector->getToLane() == detector->getFromLane() ){
					QString stationIdStr = detector->getExternalId().split( "_" ).first();
					bool ok;
					int stationId = stationIdStr.toInt( &ok );
					if( ok ){
						DetectorStats * stats = mDetectors.value( stationId );
						if( stats == NULL ){
							stats = new DetectorStats();
							mDetectors[stationId] = stats;
							stats->idDetector = stationId;
						}
						stats->detectors.append( detector );
					}
				}
			}
		}
	}
}

void INFLOExtension::postManage( double time )
{
	mCurrentTime += cycle * 1000;
	double simDuration = replication->getExperiment()->getScenario()->getDuration().toSeconds();
	bool lastInterval = GK::AreEqual( time, simDuration, cycle ) == true;

	manageDetectors();
	if( mCurrentTime >= mNextInfloStats && lastInterval == false ){
		mNextInfloStats += INFLOFreq * 1000;
		doTheManage( AKIGetTimeSta() );
		mInterval++;
	}
	if( mApplySpeeds ){
		manageSpeedAction( AKIGetTimeSta() );
	}
}

void INFLOExtension::manageSpeedAction( double timeSta )
{
	QMap< double, QVector<SpeedActionInfo> >::iterator iter = mDelayedActions.begin();
	GKTimeDuration time = GKTimeDuration(0,0,0).addSecs( timeSta );

	for( ; iter != mDelayedActions.end(); ){
		if( iter.key() < timeSta ){
			foreach( SpeedActionInfo actionInfo, iter.value() ){
				removeActionsFromSection(  actionInfo.idSection );
				void * action = AKIActionAddDetailedSpeedAction( actionInfo.idSection,actionInfo. laneId, actionInfo.segment, actionInfo.speed, actionInfo.vehPos, actionInfo.compliance );
				mActions[actionInfo.idSection].append( action );

				qDebug("Time: %s. Speed change on section %d (from MM %f) segment %d: new speed %f",time.toString(), actionInfo.idSection,
											  actionInfo.mileMarker , actionInfo.segment, actionInfo.speed );
			}
			iter.value().clear();
			iter = mDelayedActions.erase( iter );
		}else{
			iter++;
		}
	}
}

void INFLOExtension::manageDetectors()
{
	foreach( DetectorStats* stats, mDetectors ){
		foreach( GKDetector * detector, stats->detectors ){
			int count = AKIDetGetCounterCyclebyId( detector->getId(), 0 );
			if( count > 0 ){
				stats->counts += count;
				double speed = AKIDetGetSpeedCyclebyId( detector->getId(), 0 );
				if( speed > 0 ){
					stats->speed.addValue( speed*count );
				}
				double occupancy = AKIDetGetTimeOccupedCyclebyId( detector->getId(), 0 );
				if( occupancy >= 0 ){
					stats->occupancy += occupancy;
				}
			}
		}
	}
}

QVector<INFLOExtension::DetectorInfo> INFLOExtension::collectDetectorStats() const
{
	QVector<INFLOExtension::DetectorInfo> res;

	foreach( DetectorStats* stats, mDetectors ){
		INFLOExtension::DetectorInfo info;
		info.detectorId = stats->idDetector;
		info.realTime = mCurrentTime;
		info.volume = stats->counts;
		info.occupancy = std::min<double>( 100, (stats->occupancy/stats->detectors.size()) / INFLOFreq );
		info.speed = stats->speed.getMean( stats->counts );
		stats->counts = 0;
		stats->occupancy = 0;
		stats->speed.reset();
		res.append( info );
	}
	return res;
}

QVector<INFLOExtension::VehicleInfo> INFLOExtension::collectVehicleStats()
{
	GKColumn * endMileMarkerColumn = mModel->getColumn( "GKSection::EndMileMarker");
	QVector<INFLOExtension::VehicleInfo> res;
	int cacc1 = AKIVehGetVehTypeInternalPosition( mCacc1 );
	int cacc2 = AKIVehGetVehTypeInternalPosition( mCacc2 );
	foreach( GKSection * section, mSubpath->getRoute() ){
		int nbVehicles = AKIVehStateGetNbVehiclesSection( section->getId(), true );
		for( int vehicleIndex = 0; vehicleIndex < nbVehicles; vehicleIndex++ ){
			InfVeh infVeh = AKIVehStateGetVehicleInfSection( section->getId(), vehicleIndex );
			if( infVeh.type == cacc1 || infVeh.type == cacc2 ){
				if( packageDropped() == false ){
					INFLOExtension::VehicleInfo vehicleInfo;
					vehicleInfo.idVeh = infVeh.idVeh;
					vehicleInfo.speed = infVeh.CurrentSpeed;
					vehicleInfo.idSection = section->getId();
					vehicleInfo.position = infVeh.CurrentPos * feetToMile;// #feet to miles
					vehicleInfo.queued = infVeh.stopped;
					vehicleInfo.mileMarker = std::min<double>( vehicleInfo.position + mSectionMileMarkers[section->getId()],
							section->getDataValueDouble( endMileMarkerColumn ) );
					res.append( vehicleInfo );
				}else{
					mNbPackageDropped++;
				}
			}
		}
	}

	return res;
}

bool INFLOExtension::packageDropped() const
{
	bool res = false;

	double randomValue = mRandomBall->uniform( 0.0, 1.0 );
	res = randomValue < mPackageLoss;
	return res;
}

void INFLOExtension::writeData( const QVector<INFLOExtension::DetectorInfo> & detectorsInfo, const QVector<INFLOExtension::VehicleInfo>& vehiclesInfo )
{
	GKDataBaseInfo dbinfo;
	dbinfo.setDatabaseName( mDbLocation );
	dbinfo.setDriverName( "Access" );
	{
		QSqlDatabase db = GKAddDatabase( NULL, &dbinfo, "INFLO" );
		if( db.isValid() ){
			if( db.open() ){
				QSqlQuery query( db );
				db.transaction();
				QString cmd;
				foreach( INFLOExtension::DetectorInfo detInfo, detectorsInfo ){
					cmd = QString( "INSERT INTO TME_TSSData_Input (DSId, Volume, Occupancy, AvgSpeed) VALUES (%1,%2,%3,%4);" ).arg( detInfo.detectorId ).arg(detInfo.volume).arg(detInfo.occupancy).arg(detInfo.speed);
					if ( query.exec(cmd) == false ) {
						qDebug() << query.lastError().text();
					}
				}
				foreach( INFLOExtension::VehicleInfo vehInfo, vehiclesInfo ){
					cmd = QString( "INSERT INTO TME_CVData_Input (NomadicDeviceID, Speed, MMLocation, CVQueuedState) VALUES (%1,%2,%3,%4);" ).arg(vehInfo.idVeh).arg(vehInfo.speed).arg(vehInfo.mileMarker).arg(vehInfo.queued);
					if ( query.exec(cmd) == false ) {
						qDebug() << query.lastError().text();
					}
				}

				db.commit();
			}else{
				qDebug() << db.lastError().text();
			}
			db.close();
		}
	}
	QSqlDatabase::removeDatabase( "INFLO" );

	QFile file( mSyncFile );
	file.open( QIODevice::WriteOnly );
	file.write( "TSSData\nCVData\n" );
	file.close();
}

QVector< INFLOExtension::SpeedInfo > INFLOExtension::getSpeed() const
{
	QVector< INFLOExtension::SpeedInfo > res;
	GKDataBaseInfo dbinfo;
	dbinfo.setDatabaseName( mDbLocation );
	dbinfo.setDriverName( "access" );
	{
		QSqlDatabase db = GKAddDatabase( NULL, &dbinfo, "INFLO" );
		if( db.isValid() ){
			if( db.open() ){
				QSqlQuery query( db );
				qDebug() << "Reading DB";
				QString cmd;
				cmd = "SELECT TMEOutput_SPDHARMMessage_CV.* FROM TMEOutput_SPDHARMMessage_CV;";
				query.setForwardOnly( true );
				if( query.exec( cmd ) ){
					while( query.next() ){
						INFLOExtension::SpeedInfo info;
						info.speed = query.value( 2 ).toDouble();
						info.endMM = query.value( 4 ).toDouble();

						qDebug() << "MM " << info.endMM << " Speed " << info.speed;
						res.append( info );
					}
				}

				db.transaction();
				cmd = "DELETE * FROM TMEOutput_SPDHARMMessage_CV";
				if( query.exec(cmd) == false ){
					qDebug() << "failed to delete";
				}
				db.commit();
			}
			db.close();
		}
	}
	QSqlDatabase::removeDatabase( "INFLO" );

	return res;
}

double INFLOExtension::findSpeedToApply( const QVector< INFLOExtension::SpeedInfo >& speedData, double mileMarkerFrom, double mileMarkerTo )
{
	double res = DBL_MAX;

	QVector< INFLOExtension::SpeedInfo >::const_iterator iter, iterEnd;
	iter = speedData.begin();
	iterEnd = speedData.end();

	for( ; iter != iterEnd; ++iter ){
		double fromMM = iter->endMM - 0.1;
		if( mileMarkerTo >= fromMM && mileMarkerFrom <= iter->endMM ){
			res = std::min<double>( res, iter->speed );
		}
	}

	return res;
}

void INFLOExtension::updateSD( const QVector< INFLOExtension::SpeedInfo >& speedData, double timeSta )
{
	QMap< double, MileMarkerInfo >::const_iterator itSectionSegments, itSectionSegmentsEnd;
	itSectionSegments = mMileMarkerInfo.begin();
	itSectionSegmentsEnd = mMileMarkerInfo.end();

	for( ;itSectionSegments != itSectionSegmentsEnd; ++itSectionSegments ){
		double speed = findSpeedToApply( speedData, itSectionSegments.value().startMM, itSectionSegments.key() );
		if( speed != DBL_MAX && GK::AreEqual( speed, 0) == false ){
			int laneId = -1;
			int segment = itSectionSegments.value().segment;
			double complianceLevel = 1.0 - mPackageLoss;

			SpeedActionInfo infoCacc1;
			infoCacc1.vehPos = AKIVehGetVehTypeInternalPosition( mCacc1 );
			infoCacc1.laneId = laneId;
			infoCacc1.segment = segment;
			infoCacc1.compliance = complianceLevel;
			infoCacc1.idSection = itSectionSegments.value().section->getId();
			infoCacc1.speed = speed;
			infoCacc1.mileMarker = itSectionSegments.key();
			mDelayedActions[timeSta+mLatency].append( infoCacc1 );

			SpeedActionInfo infoCacc2;
			infoCacc2.vehPos = AKIVehGetVehTypeInternalPosition( mCacc2 );
			infoCacc2.laneId = laneId;
			infoCacc2.segment = segment;
			infoCacc2.compliance = complianceLevel;
			infoCacc2.idSection = itSectionSegments.value().section->getId();
			infoCacc2.speed = speed;
			infoCacc2.mileMarker = itSectionSegments.key();
			mDelayedActions[timeSta+mLatency].append( infoCacc2 );
		}
	}
}

void INFLOExtension::removeActionsFromSection( int sectionId )
{
	foreach( void* action, mActions.value( sectionId ) ){
		AKIActionRemoveAction( action );
	}
	mActions.remove(  sectionId );
}

INFLOExtension::MileMarkerInfo INFLOExtension::findMileMarkerInfo( double endMileMarker ) const
{
	QMap< double, MileMarkerInfo >::const_iterator iter, iterEnd;
	MileMarkerInfo res;
	res.section = NULL;
	iter = mMileMarkerInfo.begin();
	iterEnd = mMileMarkerInfo.end();
	bool found = false;
	double tolerance = 0.1;
	for( ; iter != iterEnd && found == false; ++iter ){
		double startMM = iter.value().startMM;
		double endMM = iter.key();
		if( iter.value().startMM <= endMileMarker && ( endMileMarker < iter.key() || fabs( endMileMarker - iter.key() ) < tolerance ) ){
			res.section = iter.value().section;
			res.segment = iter.value().segment;
			res.startMM = iter.value().startMM;
			found = true;
		}
	}
	return res;
}

void INFLOExtension::doTheManage( double timeSta )
{
	QVector<INFLOExtension::DetectorInfo> detectorInfo = collectDetectorStats();
	QVector<INFLOExtension::VehicleInfo> vehicleInfo = collectVehicleStats();
	//writes the data into the DB
	writeData( detectorInfo, vehicleInfo );
	//#waits for INFLO to compute the new speeds
	Sleep( INFLOSleep * 1000 );
	QVector< INFLOExtension::SpeedInfo > speedData = getSpeed();
	updateSD( speedData, timeSta );
	if( mApplySpeeds == false ){
		qDebug() << "INFLO speeds won't be applied in Aimsun Model";
	}
}



void INFLOExtension::finish()
{
	qDebug() << "Number of packet loss (vehicles not sending their position): " << mNbPackageDropped;
	qDeleteAll( mDetectors );
}


setupINFLODB.py
import os

import win32com.client as com

class DetectorInfo:
	def __init__(self):
		self.detectorId = 0
		self.stationId = 0
		self.sectionId = 0
		self.mileMarker = 0
		self.nbLanes = 0		
		
def export( subpath ):
	global db	
	global detectorsInfo
	global linksInfo
	global mileMarkerColumn
	global endMileMarkerColumn
	global toMiles	
	
	column = subpath.getModel().getColumn("GKSubpath::INFLO::IncreasingDirection")
	conn = com.Dispatch(r'ADODB.Connection')	
	DSN = ('PROVIDER = Microsoft.ACE.OLEDB.12.0;Data Source = ' + db)
	conn.Open(DSN)
	rs = com.Dispatch(r'ADODB.Recordset')
	state1 = "DELETE * FROM Configuration_TSSDetectorStation"
	state2 = "DELETE * FROM Configuration_RoadwaySubLinks"
	state3 = "DELETE * FROM Configuration_Roadway"
	state4 = "DELETE * FROM Configuration_RoadwayLinks"
	state5 = "DELETE * FROM Configuration_TSSDetectionZone"
	rs = conn.Execute(state1) #clear detector station sheet
	rs = conn.Execute(state2) #clear roadway sublinks
	rs = conn.Execute(state3) #clear roadway 
	rs = conn.Execute(state4) #clear roadway links
	rs = conn.Execute(state5) #clear detector zones
	
	mileMarker = 0
	indexSublink = 1
	indexZone = 1
	for section in subpath.getRoute():				
		section.setDataValue( mileMarkerColumn, QVariant(mileMarker) )
		sectionLength = max( round(section.length2D() * toMiles, 1), 0.1 )
		sectionLengthMM = mileMarker + sectionLength
		section.setDataValue( endMileMarkerColumn, QVariant(sectionLengthMM) )
		nbDetectors = 0
		sql = ""
		detectorStationId = ""
		if section in detectorsInfo:
			nbDetectors = len(detectorsInfo[section])
			if nbDetectors > 0:
				detectorStationId = ""
				for indexDetector in range(nbDetectors):
					separator = ""
					if indexDetector > 0:
						separator = " "
					detectorStationId = str(detectorsInfo[section][indexDetector].stationId)
					nbDetectors = 1
			sql = "INSERT INTO Configuration_RoadwayLinks (RoadwayId, LinkId, BeginMM, EndMM, NumberLanes, SpeedLimit, Direction, NumberDetectorStations, DetectorStations) VALUES (" +\
				str(subpath.getId()) + "," + str(section.getId()) + "," + str(mileMarker) + "," + str(sectionLengthMM) + "," + str(section.getDataValueIntByID(GKSection.nblanesAtt) ) + "," +\
				str(model.speedModelToGui( section.getSpeed() ) ) + ",\"" + str(subpath.getExternalId()) + "\"," + str(nbDetectors) + ",\"" + detectorStationId + "\");"	
		else:
			sql = "INSERT INTO Configuration_RoadwayLinks (RoadwayId, LinkId, BeginMM, EndMM, NumberLanes, SpeedLimit, Direction, NumberDetectorStations) VALUES (" +\
				str(subpath.getId()) + "," + str(section.getId()) + "," + str(mileMarker) + "," + str(sectionLengthMM) + "," + str(section.getDataValueIntByID(GKSection.nblanesAtt) ) + "," +\
				str(model.speedModelToGui( section.getSpeed() ) ) + ",\"" + str(subpath.getExternalId()) + "\"," + str(nbDetectors) + ");"
		print sql
		rs = conn.Execute( sql )	
		# nbSegments = section.getNbSegments()
		
		# for index in range( nbSegments ):	
			# segmentLength = section.getSegmentLength( index ) * toMiles
			# sql = "INSERT INTO Configuration_RoadwaySubLinks (RoadwayId, SubLinkId, BeginMM, EndMM, Direction, SpeedLimit, NumberLanes) VALUES (" +\
			# str(subpath.getId()) + "," + str(indexSublink) + "," + str(mileMarker) + "," + str(mileMarker+segmentLength) + ",\"" + str(subpath.getExternalId()) + "\"," +str(model.speedModelToGui( section.getSpeed() ) ) + "," + \
			# str(section.getDataValueIntByID(GKSection.nblanesAtt) ) + ");"	
			# indexSublink = indexSublink + 1	
			# print sql
			# rs = conn.Execute( sql )
			# mileMarker = mileMarker + segmentLength
		print sectionLength*10
		for index in range( int(sectionLength*10) ):
			segmentLength = 1
			sql = "INSERT INTO Configuration_RoadwaySubLinks (RoadwayId, SubLinkId, BeginMM, EndMM, Direction, SpeedLimit, NumberLanes) VALUES (" +\
			 str(subpath.getId()) + "," + str(indexSublink) + "," + str(mileMarker) + "," + str(mileMarker + 0.1) + ",\"" + str(subpath.getExternalId()) + "\"," +str(model.speedModelToGui( section.getSpeed() ) ) + "," + \
			 str(section.getDataValueIntByID(GKSection.nblanesAtt) ) + ");"
			indexSublink = indexSublink + 1
			mileMarker = mileMarker + 0.1
			rs = conn.Execute( sql )
		mileMarker = sectionLengthMM
		
		#detectors in section 	
		if section in detectorsInfo:
			print str(section.getId())
			detectionZones = ""
			stationId = ""
			detMileMarker = 0
			for detInfo in detectorsInfo[section]:	
				stationId = detInfo.stationId
				detMileMarker = detInfo.mileMarker 
				for lane in range(1, detInfo.nbLanes+1 ):
					sql = "INSERT INTO Configuration_TSSDetectionZone (DSId, DZId, Direction, LaneNumber) VALUES (" + \
					str(detInfo.detectorId) + "," + str( indexZone ) + ",\"" + subpath.getExternalId() + "\"," + str(lane) + ");"
					if detectionZones != "":
						detectionZones = detectionZones + ","
					detectionZones = detectionZones + str(indexZone)
					print sql
					indexZone = indexZone + 1
					rs = conn.Execute( sql )
			
			sql = "INSERT INTO Configuration_TSSDetectorStation (LinkId, DSId, MMLocation, NumberLanes, NumberDetectionZones, DetectionZones) VALUES (" + \
			str(section.getId()) + "," + str(stationId) + "," + str(round(section.getDataValueDouble(mileMarkerColumn) + detMileMarker, 1)) + \
			"," + str(len(detectorsInfo[section])) + "," + str(len(detectorsInfo[section])) + ",\"" + detectionZones + "\");"		
			
			print sql
			rs = conn.Execute( sql )
				
				
				
	sql = "INSERT INTO Configuration_Roadway (RoadwayId, Name, BeginMM, EndMM, Direction, MMIncreasingDirection, RecurringCongestionMMLocation) VALUES (" +\
			str(subpath.getId()) + ",\"" + str(subpath.getName()) + "\"," + str(0) + "," + str(mileMarker) + ",\"" + str( subpath.getExternalId() ) + "\",\"" + \
			str(subpath.getDataValueString(column)) + "\"," + str( 8 ) + ");"	
	print sql
	rs = conn.Execute( sql )	
	
	
	conn.close

def collectSubpathInfo( subpath ):
	global detectorsInfo
	global toMiles	
	for section in subpath.getRoute():
		if section.getTopObjects() != None:
			detectorsByPos = {}
			for topObject in section.getTopObjects():
				if topObject.isA( "GKDetector" ):
					detectorsByPos[topObject.getPosition()] = topObject
			
			for detector in detectorsByPos.itervalues():
				eid = "%s"%detector.getExternalId()
				if eid != "":
					stationId = int(eid.split("_")[0])
					if detector.getFromLane() == detector.getToLane() and stationId != 0:
						info = DetectorInfo()
						info.detectorId = detector.getId()
						info.section = detector.getSection()
						info.stationId = stationId
						info.mileMarker = detector.getPosition() * toMiles
						info.nbLanes = detector.getToLane() - detector.getFromLane() + 1
						if detector.getSection() not in detectorsInfo:
							detectorsInfo[detector.getSection()] = []
						detectorsInfo[detector.getSection()].append( info )

def initModel():
	global mileMarkerColumn
	global endMileMarkerColumn
	sectionType = model.getType("GKSection")
	mileMarkerColumn = sectionType.addColumn( "GKSection::MileMarker", "Start Mile Marker", GKColumn.Double )
	endMileMarkerColumn = sectionType.addColumn( "GKSection::EndMileMarker", "End Mile Marker", GKColumn.Double )
	
def main():
	initModel()
	collectSubpathInfo( target )
	export( target )

toMiles = 0.000621371
mileMarkerColumn = None
endMileMarkerColumn = None
detectorsInfo = {}
db = "C:/Users/paolo/Desktop/INFLO PM4/INFLODatabase.accdb"
main()
print "done"